File: typescript/worldbank-gdp.ts
// To parse this data:
//
//   import { Convert } from "./file";
//
//   const gdp = Convert.toGdp(json);
//
// These functions will throw an error if the JSON doesn't
// match the expected interface, even if the JSON is valid.

export interface GDPElement {
    indicator: Country;
    country:   Country;
    value:     string;
    decimal:   string;
    date:      string;
}

export interface Country {
    id:    ID;
    value: Value;
}

export enum ID {
    CN = "CN",
    NyGdpMktpCD = "NY.GDP.MKTP.CD",
    Us = "US",
}

export enum Value {
    China = "China",
    GDPCurrentUS = "GDP (current US$)",
    UnitedStates = "United States",
}

export interface PurpleGDP {
    page:     number;
    pages:    number;
    per_page: string;
    total:    number;
}

// Converts JSON strings to/from your types
// and asserts the results of JSON.parse at runtime
export namespace Convert {
    export function toGdp(json: string): Array<GDPElement[] | PurpleGDP> {
        return cast(JSON.parse(json), a(u(a(r("GDPElement")), r("PurpleGDP"))));
    }

    export function gdpToJson(value: Array<GDPElement[] | PurpleGDP>): string {
        return JSON.stringify(uncast(value, a(u(a(r("GDPElement")), r("PurpleGDP")))), null, 2);
    }

    function invalidValue(typ: any, val: any): never {
        throw Error(`Invalid value ${JSON.stringify(val)} for type ${JSON.stringify(typ)}`);
    }

    function jsonToJSProps(typ: any): any {
        if (typ.jsonToJS === undefined) {
            var map: any = {};
            typ.props.forEach((p: any) => map[p.json] = { key: p.js, typ: p.typ });
            typ.jsonToJS = map;
        }
        return typ.jsonToJS;
    }

    function jsToJSONProps(typ: any): any {
        if (typ.jsToJSON === undefined) {
            var map: any = {};
            typ.props.forEach((p: any) => map[p.js] = { key: p.json, typ: p.typ });
            typ.jsToJSON = map;
        }
        return typ.jsToJSON;
    }

    function transform(val: any, typ: any, getProps: any): any {
        function transformPrimitive(typ: string, val: any): any {
            if (typeof typ === typeof val) return val;
            return invalidValue(typ, val);
        }

        function transformUnion(typs: any[], val: any): any {
            // val must validate against one typ in typs
            var l = typs.length;
            for (var i = 0; i < l; i++) {
                var typ = typs[i];
                try {
                    return transform(val, typ, getProps);
                } catch (_) {}
            }
            return invalidValue(typs, val);
        }

        function transformEnum(cases: string[], val: any): any {
            if (cases.indexOf(val) !== -1) return val;
            return invalidValue(cases, val);
        }

        function transformArray(typ: any, val: any): any {
            // val must be an array with no invalid elements
            if (!Array.isArray(val)) return invalidValue("array", val);
            return val.map(el => transform(el, typ, getProps));
        }

        function transformObject(props: { [k: string]: any }, additional: any, val: any): any {
            if (val === null || typeof val !== "object" || Array.isArray(val)) {
                return invalidValue("object", val);
            }
            var result: any = {};
            Object.getOwnPropertyNames(props).forEach(key => {
                const prop = props[key];
                const v = Object.prototype.hasOwnProperty.call(val, key) ? val[key] : undefined;
                result[prop.key] = transform(v, prop.typ, getProps);
            });
            Object.getOwnPropertyNames(val).forEach(key => {
                if (!Object.prototype.hasOwnProperty.call(props, key)) {
                    result[key] = transform(val[key], additional, getProps);
                }
            });
            return result;
        }

        if (typ === "any") return val;
        if (typ === null) {
            if (val === null) return val;
            return invalidValue(typ, val);
        }
        if (typ === false) return invalidValue(typ, val);
        while (typeof typ === "object" && typ.ref !== undefined) {
            typ = typeMap[typ.ref];
        }
        if (Array.isArray(typ)) return transformEnum(typ, val);
        if (typeof typ === "object") {
            return typ.hasOwnProperty("unionMembers") ? transformUnion(typ.unionMembers, val)
                : typ.hasOwnProperty("arrayItems")    ? transformArray(typ.arrayItems, val)
                : typ.hasOwnProperty("props")         ? transformObject(getProps(typ), typ.additional, val)
                : invalidValue(typ, val);
        }
        return transformPrimitive(typ, val);
    }

    function cast<T>(val: any, typ: any): T {
        return transform(val, typ, jsonToJSProps);
    }

    function uncast<T>(val: T, typ: any): any {
        return transform(val, typ, jsToJSONProps);
    }

    function a(typ: any) {
        return { arrayItems: typ };
    }

    function u(...typs: any[]) {
        return { unionMembers: typs };
    }

    function o(props: any[], additional: any) {
        return { props, additional };
    }

    function m(additional: any) {
        return { props: [], additional };
    }

    function r(name: string) {
        return { ref: name };
    }

    const typeMap: any = {
        "GDPElement": o([
            { json: "indicator", js: "indicator", typ: r("Country") },
            { json: "country", js: "country", typ: r("Country") },
            { json: "value", js: "value", typ: "" },
            { json: "decimal", js: "decimal", typ: "" },
            { json: "date", js: "date", typ: "" },
        ], false),
        "Country": o([
            { json: "id", js: "id", typ: r("ID") },
            { json: "value", js: "value", typ: r("Value") },
        ], false),
        "PurpleGDP": o([
            { json: "page", js: "page", typ: 0 },
            { json: "pages", js: "pages", typ: 0 },
            { json: "per_page", js: "per_page", typ: "" },
            { json: "total", js: "total", typ: 0 },
        ], false),
        "ID": [
            "CN",
            "NY.GDP.MKTP.CD",
            "US",
        ],
        "Value": [
            "China",
            "GDP (current US$)",
            "United States",
        ],
    };
}

File: python/worldbank-gdp.py
# To use this code, make sure you
#
#     import json
#
# and then, to convert JSON from a string, do
#
#     result = gdp_from_dict(json.loads(json_string))

from enum import Enum
from typing import Any, List, Union, TypeVar, Type, cast, Callable


T = TypeVar("T")
EnumT = TypeVar("EnumT", bound=Enum)


def to_enum(c: Type[EnumT], x: Any) -> EnumT:
    assert isinstance(x, c)
    return x.value


def from_str(x: Any) -> str:
    assert isinstance(x, str)
    return x


def to_class(c: Type[T], x: Any) -> dict:
    assert isinstance(x, c)
    return cast(Any, x).to_dict()


def from_int(x: Any) -> int:
    assert isinstance(x, int) and not isinstance(x, bool)
    return x


def from_list(f: Callable[[Any], T], x: Any) -> List[T]:
    assert isinstance(x, list)
    return [f(y) for y in x]


def from_union(fs, x):
    for f in fs:
        try:
            return f(x)
        except:
            pass
    assert False


class ID(Enum):
    CN = "CN"
    NY_GDP_MKTP_CD = "NY.GDP.MKTP.CD"
    US = "US"


class Value(Enum):
    CHINA = "China"
    GDP_CURRENT_US = "GDP (current US$)"
    UNITED_STATES = "United States"


class Country:
    id: ID
    value: Value

    def __init__(self, id: ID, value: Value) -> None:
        self.id = id
        self.value = value

    @staticmethod
    def from_dict(obj: Any) -> 'Country':
        assert isinstance(obj, dict)
        id = ID(obj.get("id"))
        value = Value(obj.get("value"))
        return Country(id, value)

    def to_dict(self) -> dict:
        result: dict = {}
        result["id"] = to_enum(ID, self.id)
        result["value"] = to_enum(Value, self.value)
        return result


class PurpleGDP:
    indicator: Country
    country: Country
    value: str
    decimal: int
    date: int

    def __init__(self, indicator: Country, country: Country, value: str, decimal: int, date: int) -> None:
        self.indicator = indicator
        self.country = country
        self.value = value
        self.decimal = decimal
        self.date = date

    @staticmethod
    def from_dict(obj: Any) -> 'PurpleGDP':
        assert isinstance(obj, dict)
        indicator = Country.from_dict(obj.get("indicator"))
        country = Country.from_dict(obj.get("country"))
        value = from_str(obj.get("value"))
        decimal = int(from_str(obj.get("decimal")))
        date = int(from_str(obj.get("date")))
        return PurpleGDP(indicator, country, value, decimal, date)

    def to_dict(self) -> dict:
        result: dict = {}
        result["indicator"] = to_class(Country, self.indicator)
        result["country"] = to_class(Country, self.country)
        result["value"] = from_str(self.value)
        result["decimal"] = from_str(str(self.decimal))
        result["date"] = from_str(str(self.date))
        return result


class FluffyGDP:
    page: int
    pages: int
    per_page: int
    total: int

    def __init__(self, page: int, pages: int, per_page: int, total: int) -> None:
        self.page = page
        self.pages = pages
        self.per_page = per_page
        self.total = total

    @staticmethod
    def from_dict(obj: Any) -> 'FluffyGDP':
        assert isinstance(obj, dict)
        page = from_int(obj.get("page"))
        pages = from_int(obj.get("pages"))
        per_page = int(from_str(obj.get("per_page")))
        total = from_int(obj.get("total"))
        return FluffyGDP(page, pages, per_page, total)

    def to_dict(self) -> dict:
        result: dict = {}
        result["page"] = from_int(self.page)
        result["pages"] = from_int(self.pages)
        result["per_page"] = from_str(str(self.per_page))
        result["total"] = from_int(self.total)
        return result


def gdp_from_dict(s: Any) -> List[Union[List[PurpleGDP], FluffyGDP]]:
    return from_list(lambda x: from_union([FluffyGDP.from_dict, lambda x: from_list(PurpleGDP.from_dict, x)], x), s)


def gdp_to_dict(x: List[Union[List[PurpleGDP], FluffyGDP]]) -> Any:
    return from_list(lambda x: from_union([lambda x: to_class(FluffyGDP, x), lambda x: from_list(lambda x: to_class(PurpleGDP, x), x)], x), x)
